src/server/middleware/auth.ts

```diff
@@ -176,46 +142,197 @@ export const onRequest = async ({
+      req.url === "/" ||
+      req.url === "/system/health" ||
+      req.url === "/json" ||
+      req.url.startsWith("/auth/user") ||
+      req.url.startsWith("/transaction/status")
+    ) {
+      return { isAuthed: true };
+    }
+  } else if (req.method === "POST") {
+    if (
+      req.url.startsWith("/auth/payload") ||
+      req.url.startsWith("/auth/login") ||
+      req.url.startsWith("/auth/switch-account") ||
+      req.url.startsWith("/auth/logout")
+    ) {
       return { isAuthed: true };
     }
+
+    if (req.url.startsWith("/relayer/")) {
+      const relayerId = req.url.slice("/relayer/".length);
+      if (uuidValidate(relayerId)) {
+        // "Relay transaction" endpoint which handles its own authentication.
+        return { isAuthed: true };
+      }
+    }
+  }
+
+  return { isAuthed: false };
+};
+
+/**
+ * Handle websocket request: auth via access token
+ * Allow a request that provides a non-revoked access token for an owner/admin
+ * Handle websocket auth separately.
+ * @param req FastifyRequest
+ * @param getUser
+ * @returns AuthResponse
+ * @async
+ */
+const handleWebsocketAuth = async (
+  req: FastifyRequest,
+  getUser: ReturnType<typeof ThirdwebAuth<TAuthData, TAuthSession>>["getUser"],
+): Promise<AuthResponse> => {
+  const { token: jwt } = req.query as { token: string };
+
+  const token = await getAccessToken({ jwt });
+  if (token && token.revokedAt === null) {
+    // Set as a header for `getUsers` to parse the token.
+    req.headers.authorization = `Bearer ${jwt}`;
+    const user = await getUser(req);
+    if (
+      user?.session?.permissions === Permission.Owner ||
+      user?.session?.permissions === Permission.Admin
+    ) {
+      return { isAuthed: true, user };
+    }
+  }
+
+  // Destroy the websocket connection.
+  req.raw.socket.write("HTTP/1.1 401 Unauthorized\r\n\r\n");
+  req.raw.socket.destroy();
+  return { isAuthed: false };
+};
+
+/**
+ * Auth via keypair.
+ * Allow a request that provides a JWT signed by an ES256 private key
+ * matching the configured public key.
+ * @param req FastifyRequest
+ * @returns AuthResponse
+ */
+const handleKeypairAuth = async (
+  req: FastifyRequest,
+): Promise<AuthResponse> => {
+  // The keypair auth feature must be explicitly enabled.
+  if (!env.ENABLE_KEYPAIR_AUTH) {
+    return { isAuthed: false };
+  }
+
+  const jwt = getJWT(req);
+  if (!jwt) {
+    return { isAuthed: false };
+  }
+
+  let error: string | undefined;
+  try {
+    const payload = jsonwebtoken.decode(jwt, { json: true });
+    if (!payload?.iss) {
+      error = 'The "iss" field must contain an ES256 public key.';
+      throw error;
+    }
+
+    const keypair = await getKeypair({ publicKey: payload.iss });
+    if (!keypair || keypair.publicKey !== payload.iss) {
+      error =
+        "The provided public key is not configured for this Engine instance.";
+      throw error;
+    }
+
+    const { exp, iat } = jsonwebtoken.verify(jwt, keypair.publicKey, {
```

Comment:
https://www.npmjs.com/package/jsonwebtoken#jwtverifytoken-secretorpublickey-options-callback

`issuer`, `clockTolerance`, etc. i think it checks expiration already?

Reply:
since expiresAt should be encoded in the JWT?

Reply:
This was an explicit way to check that you don't provide an expiration that's too long, to prevent a footgun where someone sets a 5-year expiration (might as well use the permanent access tokens).

If it doesn't seem necessary I can remove this.